doc/design/wallet/wallet-design.md

```diff
@@ -0,0 +1,768 @@
+
+                                          /// Get an value for retrieved wallet storage record
+                                          ///
+                                          /// #Params
+                                          /// storage_handle: opened storage handle (See open handler)
+                                          /// record_handle: retrieved record handle (See get_record handler)
+                                          ///
+                                          /// returns: record value
+                                          ///          Note that pointer lifetime the same as retrieved record lifetime
+                                          ///            (until record_free called)
+                                          ///          Note that null be returned if no value retrieved
+                                          get_record_value: Option<extern fn(storage_handle: u32,
+                                                                             record_handle: u32,
+                                                                             record_value_p: *mut *const u8,
+                                                                             record_value_len_p: *mut i32) -> ErrorCode>,
+
+                                          /// Get an tags for retrieved wallet record
+                                          ///
+                                          /// #Params
+                                          /// storage_handle: opened storage handle (See open handler)
+                                          /// record_handle: retrieved record handle (See get_record handler)
+                                          ///
+                                          /// returns: record tags as json
+                                          ///          Note that pointer lifetime the same as retrieved record lifetime
+                                          ///            (until record_free called)
+                                          ///          Note that null be returned if no tags retrieved
+                                          get_record_tags: Option<extern fn(storage_handle: u32,
+                                                                            record_handle: u32,
+                                                                            record_tags_p: *mut *const c_char) -> ErrorCode>,
+
+                                          /// Free retrieved wallet record (make retrieved record handle invalid)
+                                          ///
+                                          /// #Params
+                                          /// storage_handle: opened storage handle (See open_wallet_storage)
+                                          /// record_handle: retrieved record handle (See wallet_storage_get_wallet_record)
+                                          free_record: Option<extern fn(storage_handle: u32,
```

Comment:
There is no use for ErrorCode in free functions. You cannot do anything. It is probably the best if those functions do not have return value.

Reply:
I suggest to leve error code. It isn’t system free and allows internal handler and state validation.